### PR TITLE
Sanitizer

### PR DESCRIPTION
**Pull request by Rombur**
I'd like your opinion on the following:
 -  I have blacklisted all the tests for the leak sanitizer (which is part of the address sanitizer). The reason is that it complains about a leak from an unknown module. So there is not much I can do with that. 
 - the thread sanitizer is in beta and shows data race in every single tests. This seems strange since most of the code is not multithreaded. Since the thread sanitizer is beta right now, my plan is to not care about it right now and come back to it once it is more stable.
 - there is a bug in a Moab test that I don't understand (free of an array that was not allocated).

### CODE REVIEW

File: scripts/docker_clang_env.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+export CC=mpicc
+export CXX=mpicxx
+
+# Add clang OpenMP runtime library to LD_LIBRARY_PATH
+export LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH
+
+SANITIZER=$1
+if [ "$SANITIZER" == "undefined_sanitizer" ]
+  then FLAGS="-fsanitize=address -fsanitize=undefined -fsanitize-blacklist=/scratch/source/trilinos/release/DataTransferKit/scripts/undefined_blacklist.txt"
+  # When using the sanitizer, Fotran needs to be disable. Otherwise, we will get
```
Comment (aprokop) [inline scripts/docker_clang_env.sh]: disable -> disabled.

File: scripts/docker_clang_env.sh [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  # a runtime error, because clang and gfortran requires two different
```
Comment (aprokop) [inline scripts/docker_clang_env.sh]: requires -> require

File: packages/Adapters/ClassicDTK/test/field/tstFieldTools.cpp [hunk 3]
```diff
@@ -344,7 +344,7 @@ TEUCHOS_UNIT_TEST( FieldTools, scalar_field_test )
     }
 
     Teuchos::Array<double> q_norms;
-    int q = std::floor( ( 10 * std::rand() ) / RAND_MAX ) + 1;
+    int q = std::floor( ( 10 * (double)std::rand() ) / RAND_MAX ) + 1;
```
Comment (aprokop) [inline packages/Adapters/ClassicDTK/test/field/tstFieldTools.cpp]: just do `10. * std::rand()`

File: packages/Utils/src/DTK_SanitizerMacros.hpp [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+//---------------------------------------------------------------------------//
+/*!
+ * \file   DTK_SanitizerMacro.hpp
+ * \author Bruno Turcksin
+ * \brief  Macros to suppress sanitizer checks.
+ */
+//---------------------------------------------------------------------------//
+
+#if defined( __clang__ )
+#define NO_SANITIZE_UNDEFINED __attribute__( ( no_sanitize( "undefined" ) ) )
```
Comment (aprokop) [inline packages/Utils/src/DTK_SanitizerMacros.hpp]: `NO_SANITIZE_UNDEFINED` is really hard for me to remember. Can we change it to something else? Like `IGNORE_UNDEFINED_SANITIZE`?

File: packages/Adapters/Moab/test/tstMoabEntityLocalMap.cpp [hunk 5]
```diff
@@ -340,6 +340,7 @@ TEUCHOS_UNIT_TEST( MoabEntity, quad_4_test )
 
     // Test the mapping to reference frame.
     Teuchos::Array<double> ref_good_point( space_dim );
+    // This lines throws. It seems that it frees an internal Teuchos::Array.
```
Comment (aprokop) [inline packages/Adapters/Moab/test/tstMoabEntityLocalMap.cpp]: Update the comment to something like "This lines throws when run with -fsanitize=..."

File: scripts/docker_cmake [hunk 6]
```diff
@@ -32,6 +32,7 @@ ARGS=(
     ### PACKAGES CONFIGURATION ###
     -D Trilinos_ENABLE_ALL_PACKAGES=OFF
     -D Trilinos_ENABLE_ALL_OPTIONAL_PACKAGES=OFF
+    "${Trilinos_ENABLE_Fortran}"
```
Comment (aprokop) [inline scripts/docker_cmake]: Do `-D Trilinos_ENABLE_Fortran=${ENABLE_FORTRAN}`, and then just use `ENABLE_FORTRAN="ON"` or "OFF" in your `docker_clang_env.sh`.

File: scripts/undefined_blacklist.txt [hunk 7]
```diff
@@ -0,0 +1,3 @@
+# This two problems seem related
+src:/scratch/source/trilinos/release/packages/stk/stk_mesh/*
+src:/usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c\+\+/5.4.0/bits/*
```
Comment (aprokop) [inline scripts/undefined_blacklist.txt]: If we have instructions for building docker container (do we?), we should probably add a note to update this when the compiler version changes.

File: scripts/docker_clang_env.sh [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  # implementations of the same sanitizer library.
+  export Trilinos_ENABLE_Fortran="-DTrilinos_ENABLE_Fortran=OFF"
+  # Suppress leak sanitizer on functions we have no control of.
+  export LSAN_OPTIONS=suppressions=/scratch/source/trilinos/release/DataTransferKit/scripts/leak_blacklist.txt
+fi
```
Comment (aprokop) [inline scripts/docker_clang_env.sh]: just do `elif`

File: packages/Adapters/ClassicDTK/test/field/tstFieldTools.cpp (hunk 3 above)
Comment (Rombur) [inline packages/Adapters/ClassicDTK/test/field/tstFieldTools.cpp]: I copied what was done for the other `rands`.  I think it's better to keep it uniform in the same file.

File: packages/Adapters/Moab/test/tstMoabEntityLocalMap.cpp (hunk 5 above)
Comment (Rombur) [inline packages/Adapters/Moab/test/tstMoabEntityLocalMap.cpp]: Yes. This is just temporary until we decide what to with this line, fix it or no? I don't know the adapter for Moab, so I am not sure if it is worth spending time on this.

File: scripts/docker_cmake (hunk 6 above)
Comment (Rombur) [inline scripts/docker_cmake]: I did this so the behavior doesn't change when using `gcc`. If I do what you propose, we will need a `docker_gcc_env.sh` to set `ENABLE_FORTRAN`.

File: packages/Operators/src/SharedDomain/DTK_ConsistentInterpolationOperator.cpp [hunk 9]
```diff
@@ -187,8 +187,10 @@ void ConsistentInterpolationOperator::setupImpl(
 
             // Add a scale factor for this range entity to the scaling vector.
             DTK_CHECK( range_map->isNodeGlobalElement( range_support_ids[0] ) );
+            const double scale_factor =
+                ( domain_ids.size() != 0 ) ? 1.0 / domain_ids.size() : 1.;
             scale_vector->replaceGlobalValue( range_support_ids[0],
-                                              1.0 / domain_ids.size() );
+                                              scale_factor );
```
Comment (dalg24) [inline packages/Operators/src/SharedDomain/DTK_ConsistentInterpolationOperator.cpp]: The sanitizer does not like division by zero?

File: docker/jenkins/build-sanitizer.sh [hunk 10]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+
+# number of processes with default value
+: ${NPROC:=8}
+# cleanup workspace
+cd ${TRILINOS_DIR}/DataTransferKit
+[ -d build ] && rm -rf build
+mkdir build && cd build
+# configure trilinos with dtk
+. ../scripts/docker_clang_env.sh undefined_sanitizer
```
Comment (dalg24) [inline docker/jenkins/build-sanitizer.sh]: is that line the only difference with `docker/jenkins/build.sh`?

File: docker/jenkins/config-sanitizer.dict [hunk 11]
```diff
@@ -0,0 +1,23 @@
+{
+  'api_calls_timeout_seconds': 300,
+  'image': 'dalg24/dtk-stack',
+  'tag': 'latest',
+  'name': os.getenv("BUILD_TAG"),
+  'cmd': [
+    'sh -c "patch -d /scratch/source/trilinos/release -p1 -i /scratch/source/trilinos/release/DataTransferKit/docker/jenkins/trilinos_stk.patch"',
+    'sh -xe /scratch/source/trilinos/release/DataTransferKit/docker/jenkins/build-sanitizer.sh',
```
Comment (dalg24) [inline docker/jenkins/config-sanitizer.dict]: so here same dictionary as in `docker/jenkins/config.dict` but this line that points to the alternate build script.  is that right?

Comment (Rombur) [inline docker/jenkins/config-sanitizer.dict]: no I am still working on this

File: README.md [hunk 12]
```diff
@@ -24,7 +24,7 @@ Questions, Bug Reporting, and Issue Tracking
 --------------------------------------------
 
 Questions, bug reporting and issue tracking are provided by GitHub. Please
-report all bugs by creating a new issue with the bug tag. You can ask
+report allbugs by creating a new issue with the bug tag. You can ask
```
Comment (sslattery) [inline README.md]: allbugs -> all bugs

File: packages/Adapters/BasicGeometry/src/DTK_BasicEntitySet.hpp [hunk 13]
```diff
@@ -92,9 +93,11 @@ class BasicEntitySetIterator : public EntityIterator
     Entity *operator->(void)override;
 
     // Equal comparison operator.
+    IGNORE_UNDEFINED_SANITIZE
```
Comment (sslattery) [inline packages/Adapters/BasicGeometry/src/DTK_BasicEntitySet.hpp]: We need to make sure we have a policy of using this macro only for code that is deprecated. Any new code or code that will remain in the repo should not have this macro.

Per our previous discussions this use case is a good example of when to use the macro as the old interfaces are going to be removed.

File: packages/Adapters/Moab/test/tstMoabEntityLocalMap.cpp (hunk 5 above)
Comment (sslattery) [inline packages/Adapters/Moab/test/tstMoabEntityLocalMap.cpp]: This code will be deprecated so I am ok leaving as is.

File: scripts/docker_clang_env.sh [hunk 14]
```diff
@@ -0,0 +1,25 @@
+# Use clang instead of gcc
+export OMPI_CC=/usr/bin/clang-3.9
```
Comment (sslattery) [inline scripts/docker_clang_env.sh]: Is this hard coded for the docker container?

Comment (sslattery) [inline scripts/docker_clang_env.sh]: couldn't this entire header file instead be set up with CMake?

Comment (sslattery) [inline scripts/docker_clang_env.sh]: never mind - didn't see it was in docker_clang_env.sh ;)

File: docker/jenkins/config.dict [hunk 15]
```diff
@@ -4,15 +4,16 @@
   'tag': 'latest',
   'name': os.getenv("BUILD_TAG"),
   'cmd': [
-    'sh -c "patch -d /scratch/source/trilinos/release -p1 -i /scratch/source/trilinos/release/DataTransferKit/docker/jenkins/trilinos_stk.patch"',
-    'sh -xe /scratch/source/trilinos/release/DataTransferKit/docker/jenkins/build.sh',
+    'bash -c "patch -d /scratch/source/trilinos/release -p1 -i /scratch/source/trilinos/release/DataTransferKit/docker/jenkins/trilinos_stk.patch"',
+    'bash -xe /scratch/source/trilinos/release/DataTransferKit/docker/jenkins/build.sh',
```
Comment (dalg24) [inline docker/jenkins/config.dict]: did you find out if the `-e` options is working (abort script at first error)?

File: scripts/docker_cmake [hunk 16]
```diff
@@ -24,7 +24,7 @@ ARGS=(
     -D TPL_ENABLE_Libmesh=ON
         -D Libmesh_INCLUDE_DIRS=$LIBMESH_DIR/include
         -D Libmesh_LIBRARY_DIRS=$LIBMESH_DIR/lib
-    -D TPL_ENABLE_MOAB=ON
+    -D TPL_ENABLE_MOAB=${ENABLE_MOAB}
```
Comment (dalg24) [inline scripts/docker_cmake]: don't really like that we have to do that...  could't we just pass extra flags `-D DataTransferKit_ENABLE_MOABAdapters=OFF` or so?

File: docker/jenkins/config.dict (hunk 15 above)
Comment (Rombur) [inline docker/jenkins/config.dict]: No I didn't look

File: scripts/docker_cmake (hunk 16 above)
Comment (Rombur) [inline scripts/docker_cmake]: I can try but I just want to point out that if `ENABLE_MOAB` is empty, moab will be turned on. So the behavior of `docker_cmake` hasn't changed.

Comment (Rombur) [inline scripts/docker_cmake]: @dalg24 passing `-DDataTransferKit_ENABLE_MoabAdapters=OFF` as an extra flag doesn't work.

Comment (dalg24) [inline scripts/docker_cmake]: does CMake pick the flag up?  I made it up, not sure this is the actual flag.

File: scripts/docker_clang_env.sh [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+export OMPI_CXX=/usr/bin/clang++-3.9
+export CC=mpicc
+export CXX=mpicxx
+
+# Add clang OpenMP runtime library to LD_LIBRARY_PATH
+export LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH
+
+SANITIZER=$1
```
Comment (dalg24) [inline scripts/docker_clang_env.sh]: Why don't you use the `SANITIZE={undefined,thread}` env variable here?  It is defined in the the script `docker/jenkins/build.sh` that calls `docker_clang_env.sh`.   No need to change, I'm just trying to understand why you chose to do things this way.

File: docker/jenkins/config.dict (hunk 15 above)
Comment (dalg24) [inline docker/jenkins/config.dict]: did you eventually find out this one?

File: scripts/undefined_blacklist.txt [hunk 18]
```diff
@@ -0,0 +1,3 @@
+# This two problems seem related
```
Comment (dalg24) [inline scripts/undefined_blacklist.txt]: what do you mean by that?

Comment (Rombur) [inline scripts/undefined_blacklist.txt]: I need to blacklist `stl_vector` because of something from `stk_mesh`

**Comment by Rombur** [comment]
@dalg24 _ping_

File: docker/jenkins/config.dict (hunk 15 above)
Comment (dalg24) [inline docker/jenkins/config.dict]: I guess we'll figure that out later...

**Pull request merged**